dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSQLVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,529 @@
+    /// More information about Azure CosmosDB NoSQL distance functions here: <see href="https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/vector-search#container-vector-policies" />.
+    /// </summary>
+    private static DistanceFunction GetDistanceFunction(string? distanceFunction, string vectorPropertyName)
+    {
+        return distanceFunction switch
+        {
+            SKDistanceFunction.CosineSimilarity => DistanceFunction.Cosine,
+            SKDistanceFunction.DotProductSimilarity => DistanceFunction.DotProduct,
+            SKDistanceFunction.EuclideanDistance => DistanceFunction.Euclidean,
+            _ => throw new InvalidOperationException($"Distance function '{distanceFunction}' for {nameof(VectorStoreRecordVectorProperty)} '{vectorPropertyName}' is not supported by the Azure CosmosDB NoSQL VectorStore.")
+        };
+    }
+
+    /// <summary>
+    /// Returns <see cref="VectorDataType"/> based on vector property type.
+    /// </summary>
+    private static VectorDataType GetDataType(Type vectorDataType, string vectorPropertyName)
+    {
+        return vectorDataType switch
+        {
+#if NET5_0_OR_GREATER
+            Type type when type == typeof(ReadOnlyMemory<Half>) || type == typeof(ReadOnlyMemory<Half>?) => VectorDataType.Float16,
+#endif
+            Type type when type == typeof(ReadOnlyMemory<float>) || type == typeof(ReadOnlyMemory<float>?) => VectorDataType.Float32,
+            Type type when type == typeof(ReadOnlyMemory<byte>) || type == typeof(ReadOnlyMemory<byte>?) => VectorDataType.Uint8,
+            Type type when type == typeof(ReadOnlyMemory<sbyte>) || type == typeof(ReadOnlyMemory<sbyte>?) => VectorDataType.Int8,
+            _ => throw new InvalidOperationException($"Data type '{vectorDataType}' for {nameof(VectorStoreRecordVectorProperty)} '{vectorPropertyName}' is not supported by the Azure CosmosDB NoSQL VectorStore.")
+        };
+    }
+
+    private QueryDefinition GetSelectQuery(List<string> keys, List<string> fields)
+    {
+        var selectFields = fields.Select(field => $"x.{field}");
+        var ids = string.Join(",", keys.Select((id, index) => $"@id{index}"));
+
+        var whereClause = keys.Count == 1 ?
+            $"WHERE x.{this._keyStoragePropertyName} = @id0" :
+            $"WHERE x.{this._keyStoragePropertyName} IN ({ids})";
+
+        var query = $"SELECT {string.Join(",", selectFields)} FROM x {whereClause}";
+
+        var queryDefinition = new QueryDefinition(query);
+
+        for (var i = 0; i < keys.Count; i++)
+        {
+            queryDefinition.WithParameter($"@id{i}", keys[i]);
+        }
+
+        return queryDefinition;
+    }
+
+    private async IAsyncEnumerable<JsonObject> GetItemsAsync(QueryDefinition queryDefinition, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var iterator = this._database
+            .GetContainer(this.CollectionName)
+            .GetItemQueryIterator<JsonObject>(queryDefinition);
```

Comment:
Another option here could be to use `GetItemQueryStreamIterator`.  It will return you streams instead of JsonObjects.  This means, that you can control the deserialization of the stream yourself.
This can make things a little bit simpler in that the user doesn't need to pass the same serializer options to both the client and the connector, since the client one wouldn't actually be used.
You'd also have to use `UpsertItemStreamAsync` for upsert though.

Reply:
`JsonObject` works good in this case, since we need to manipulate with `id` and `partition key` fields, and with `JsonObject` we have easy access to them, by relying on NoSQL SDK {de}serialization. In case of `Stream` we will need to put {de}serialization logic on our side and still work with `JsonObject` to get access to these fields or transform it to `TRecord` and use reflection.

I can create a task to investigate `Stream` approach separately, if you think it will work better in long-term. Please let me know.

Reply:
Based on internal conversation, I tried the approach with `Stream`, but I still need to handle Key property serialization separately, since `id` is a reserved name for document identifier in Azure Cosmos DB NoSQL. So, even if we will operate with `Stream` objects at the end, I'm not sure it will be possible to avoid using `JsonObject`, and in this case serializer options still should be provided. I'm happy to discuss it further in case I miss something, please let me know.